instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarPathFinder.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.resources;
+
+import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+class JarPathFinder {
+  private final Supplier<String[]> getProcessHandleArguments;
+  private final Function<String, String> getSystemProperty;
+  private final Predicate<Path> fileExists;
+
+  private static Optional<Optional<Path>> jarPath = Optional.empty();
```

Comment:
Would it be possible to have `Optional<Optional<Path>>` to simplify?

Reply:
no

- the outer optional is to know if detection has happened
- the inner optional is to know if a path has been found

Reply:
> no
> 
> * the outer optional is to know if detection has happened
> * the inner optional is to know if a path has been found

This can't be quickly guess for me by reading the code.

So, `jarPath` seems to be related to 2 different things.

 Could you please update the code to make these things more explicit in the code, for example with another property related to the detection?